### ISSUE TITLE
ModuleNotFoundError: No module named 'py7zr' yet pip list shows it is...

### ISSUE THREAD

**Issue by ak2766** [royalgraphx/DarwinFetch#2]
When I execute `./DarwinFetch.ss`, I get the following output:
```
RGX Python Project Bootstrapper Version: 0.1.2
Running on Debian-based Linux...
Python 3 detected. Version: 3.12.3
ensurepip is installed.
Virtual environment '.DFetchVEnv' already exists.
Activated virtual environment: .DFetchVEnv
Checks passed...
Running DarwinFetch using python3
Traceback (most recent call last):
  File "/opt/virtualenv/DarwinFetch/src/main.py", line 12, in <module>
    import py7zr
ModuleNotFoundError: No module named 'py7zr'
```

Yet,
```
$ > python
Python 3.12.3 (main, Sep 11 2024, 14:17:37) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import py7zr
>>> 
```

Help...

**Comment by ak2766** [comment]
Should have included `pip list`:
```
$ > pip list
Package            Version
------------------ ---------
Brotli             1.1.0
certifi            2024.8.30
charset-normalizer 3.3.2
click              8.1.7
idna               3.10
inflate64          1.0.0
multivolumefile    0.2.3
pip                24.0
psutil             6.0.0
py7zr              0.22.0
pybcj              1.0.2
pycryptodomex      3.20.0
pyppmd             1.1.0
pyzstd             0.16.1
requests           2.32.3
texttable          1.7.0
tqdm               4.66.5
urllib3            2.2.3
```

**Comment by royalgraphx** [comment]
what distribution is this

**Comment by ak2766** [comment]
I'm on Ubuntu 24.04.  But I've just discovered what needs to be done.  I was already using my own virtual environment so maybe that confused it all.

I've since resolved.  After looking at the script is when I noticed that it is using its own virtualenv.  So deleted mine, and tried to run `./DarwinFetch.sh` but it still gave the same error.

In summary, the steps to run this on Ubuntu 24.04.1 after cloning the repo is:
```
. .DFetchVEnv/bin/activate
pip install -r requirements.txt
./DarwinFetch.sh
```

On my merry way now and thanks for looking into this so quickly.

**Comment by royalgraphx** [comment]
you're not supposed to do anything at all, you simply run ./DarwinFetch.sh and it will genuinely do everything for you, please refrain from **ever** specifically defining the venv and let DarwinFetch do what it needs to do.

**Comment by ak2766** [comment]
Nope.

I've just tried this in a brand spanking new LXC container and it failed with the same error:

```
Running DarwinFetch using python3
Traceback (most recent call last):
  File "/home/ubuntu/DarwinFetch/src/main.py", line 12, in <module>
    import py7zr
ModuleNotFoundError: No module named 'py7zr'
```

**Comment by royalgraphx** [comment]
i've tested the bootstrapper shell script in the supported list of distros in said script, I, have not had an issue simply git cloning, and running ./DarwinFetch.sh, thats kinda the entire point and reason why it exists

**Comment by royalgraphx** [comment]
moreover, the latest #1 commit, made 0 changes that could affect this

**Comment by ak2766** [comment]
I can record a video if you like!

**Comment by royalgraphx** [comment]
``rm -rf .DFetchVenv``
``./DarwinFetch.sh``

Make sure this is in a new terminal, where you run no previous commands

**Comment by ak2766** [comment]
In the README, it says to:
```
pip install -r requirements.txt
./DarwinFetch.sh
```
Should I do this as well?  If yes, then this might be the problem.  I never install stuff I'm testing into my main system!  I always use a virtualenv.

**Comment by ak2766** [comment]
Without the pip install..., it fails with the error I posted above...

**Comment by royalgraphx** [comment]
the README is out of date, it no longer requires a user to manually install requirements, because the script itself does it... I have no idea how your setup is, but even on a fresh Ubuntu install, this should work OOB

**Comment by ak2766** [comment]
One sec - let me capture a screenshot showing the command I'm running...

**Comment by ak2766** [comment]
Here you go.

In the output, search for `root@puma` to see all the commands I ran.

```
[Sat Sep 21 17:24:16 - 6.8.0-45-lowlatency - /dev/pts/2]
[1 ✗ (135ms) akk@akstpx1g5 10.0.0.54: ~ ]
$ > lxc launch ubuntu:noble puma
Creating puma
Starting puma                             
[Sat Sep 21 17:24:24 - 6.8.0-45-lowlatency - /dev/pts/2]
[0 ✓ (2.050s) akk@akstpx1g5 10.0.0.54: ~ ]
$ > lxc exec puma bash
root@puma:~# git clone https://github.com/royalgraphx/DarwinFetch.git
Cloning into 'DarwinFetch'...
remote: Enumerating objects: 75, done.
remote: Counting objects: 100% (75/75), done.
remote: Compressing objects: 100% (53/53), done.
remote: Total 75 (delta 35), reused 53 (delta 19), pack-reused 0 (from 0)
Receiving objects: 100% (75/75), 38.45 KiB | 1.83 MiB/s, done.
Resolving deltas: 100% (35/35), done.
root@puma:~# cd DarwinFetch/
root@puma:~/DarwinFetch# egrep clear DarwinFetch.sh 
clear
root@puma:~/DarwinFetch# sed -i '/clear/d' DarwinFetch.sh 
root@puma:~/DarwinFetch# egrep clear DarwinFetch.sh 
root@puma:~/DarwinFetch# ./DarwinFetch.sh 
RGX Python Project Bootstrapper Version: 0.1.2
Running on Debian-based Linux...
Python 3 detected. Version: 3.12.3
ensurepip is not installed: /usr/lib/python3.12/ensurepip/__init__.py is missing.
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Package python3-venv is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'python3-venv' has no installation candidate
Failed to install python3-venv
root@puma:~/DarwinFetch# apt update &>/dev/null
root@puma:~/DarwinFetch# ./DarwinFetch.sh 
RGX Python Project Bootstrapper Version: 0.1.2
Running on Debian-based Linux...
Python 3 detected. Version: 3.12.3
ensurepip is not installed: /usr/lib/python3.12/ensurepip/__init__.py is missing.
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libpython3.12-minimal libpython3.12-stdlib libpython3.12t64 python3-pip-whl python3-pkg-resources python3-setuptools
  python3-setuptools-whl python3.12 python3.12-minimal python3.12-venv
Suggested packages:
  python-setuptools-doc python3.12-doc binutils binfmt-support
The following NEW packages will be installed:
  python3-pip-whl python3-setuptools-whl python3-venv python3.12-venv
The following packages will be upgraded:
  libpython3.12-minimal libpython3.12-stdlib libpython3.12t64 python3-pkg-resources python3-setuptools python3.12
  python3.12-minimal
7 upgraded, 4 newly installed, 0 to remove and 12 not upgraded.
Need to get 11.2 MB of archives.
After this operation, 2784 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu noble-updates/main amd64 libpython3.12t64 amd64 3.12.3-1ubuntu0.2 [2339 kB]
Get:2 http://archive.ubuntu.com/ubuntu noble-updates/main amd64 python3.12 amd64 3.12.3-1ubuntu0.2 [651 kB]
Get:3 http://archive.ubuntu.com/ubuntu noble-updates/main amd64 libpython3.12-stdlib amd64 3.12.3-1ubuntu0.2 [2069 kB]
Get:4 http://archive.ubuntu.com/ubuntu noble-updates/main amd64 python3.12-minimal amd64 3.12.3-1ubuntu0.2 [2343 kB]
Get:5 http://archive.ubuntu.com/ubuntu noble-updates/main amd64 libpython3.12-minimal amd64 3.12.3-1ubuntu0.2 [834 kB]
Get:6 http://archive.ubuntu.com/ubuntu noble-updates/main amd64 python3-setuptools all 68.1.2-2ubuntu1.1 [396 kB]
Get:7 http://archive.ubuntu.com/ubuntu noble-updates/main amd64 python3-pkg-resources all 68.1.2-2ubuntu1.1 [168 kB]
Get:8 http://archive.ubuntu.com/ubuntu noble/universe amd64 python3-pip-whl all 24.0+dfsg-1ubuntu1 [1702 kB]
Get:9 http://archive.ubuntu.com/ubuntu noble-updates/universe amd64 python3-setuptools-whl all 68.1.2-2ubuntu1.1 [716 kB]
Get:10 http://archive.ubuntu.com/ubuntu noble-updates/universe amd64 python3.12-venv amd64 3.12.3-1ubuntu0.2 [5678 B]
Get:11 http://archive.ubuntu.com/ubuntu noble-updates/universe amd64 python3-venv amd64 3.12.3-0ubuntu2 [1034 B]
Fetched 11.2 MB in 4s (2771 kB/s)       
(Reading database ... 34473 files and directories currently installed.)
Preparing to unpack .../00-libpython3.12t64_3.12.3-1ubuntu0.2_amd64.deb ...
Unpacking libpython3.12t64:amd64 (3.12.3-1ubuntu0.2) over (3.12.3-1ubuntu0.1) ...
Preparing to unpack .../01-python3.12_3.12.3-1ubuntu0.2_amd64.deb ...
Unpacking python3.12 (3.12.3-1ubuntu0.2) over (3.12.3-1ubuntu0.1) ...
Preparing to unpack .../02-libpython3.12-stdlib_3.12.3-1ubuntu0.2_amd64.deb ...
Unpacking libpython3.12-stdlib:amd64 (3.12.3-1ubuntu0.2) over (3.12.3-1ubuntu0.1) ...
Preparing to unpack .../03-python3.12-minimal_3.12.3-1ubuntu0.2_amd64.deb ...
Unpacking python3.12-minimal (3.12.3-1ubuntu0.2) over (3.12.3-1ubuntu0.1) ...
Preparing to unpack .../04-libpython3.12-minimal_3.12.3-1ubuntu0.2_amd64.deb ...
Unpacking libpython3.12-minimal:amd64 (3.12.3-1ubuntu0.2) over (3.12.3-1ubuntu0.1) ...
Preparing to unpack .../05-python3-setuptools_68.1.2-2ubuntu1.1_all.deb ...
Unpacking python3-setuptools (68.1.2-2ubuntu1.1) over (68.1.2-2ubuntu1) ...
Preparing to unpack .../06-python3-pkg-resources_68.1.2-2ubuntu1.1_all.deb ...
Unpacking python3-pkg-resources (68.1.2-2ubuntu1.1) over (68.1.2-2ubuntu1) ...
Selecting previously unselected package python3-pip-whl.
Preparing to unpack .../07-python3-pip-whl_24.0+dfsg-1ubuntu1_all.deb ...
Unpacking python3-pip-whl (24.0+dfsg-1ubuntu1) ...
Selecting previously unselected package python3-setuptools-whl.
Preparing to unpack .../08-python3-setuptools-whl_68.1.2-2ubuntu1.1_all.deb ...
Unpacking python3-setuptools-whl (68.1.2-2ubuntu1.1) ...
Selecting previously unselected package python3.12-venv.
Preparing to unpack .../09-python3.12-venv_3.12.3-1ubuntu0.2_amd64.deb ...
Unpacking python3.12-venv (3.12.3-1ubuntu0.2) ...
Selecting previously unselected package python3-venv.
Preparing to unpack .../10-python3-venv_3.12.3-0ubuntu2_amd64.deb ...
Unpacking python3-venv (3.12.3-0ubuntu2) ...
Setting up python3-pkg-resources (68.1.2-2ubuntu1.1) ...
Setting up python3-setuptools-whl (68.1.2-2ubuntu1.1) ...
Setting up python3-setuptools (68.1.2-2ubuntu1.1) ...
Setting up python3-pip-whl (24.0+dfsg-1ubuntu1) ...
Setting up libpython3.12-minimal:amd64 (3.12.3-1ubuntu0.2) ...
Setting up python3.12-minimal (3.12.3-1ubuntu0.2) ...
Setting up libpython3.12-stdlib:amd64 (3.12.3-1ubuntu0.2) ...
Setting up python3.12 (3.12.3-1ubuntu0.2) ...
Setting up libpython3.12t64:amd64 (3.12.3-1ubuntu0.2) ...
Setting up python3.12-venv (3.12.3-1ubuntu0.2) ...
Setting up python3-venv (3.12.3-0ubuntu2) ...
Processing triggers for systemd (255.4-1ubuntu8.4) ...
Processing triggers for man-db (2.12.0-4build2) ...
Processing triggers for libc-bin (2.39-0ubuntu8.3) ...
Scanning processes...                                                                                                       
Scanning candidates...                                                                                                      

Restarting services...

Service restarts being deferred:
 systemctl restart unattended-upgrades.service

No containers need to be restarted.

No user sessions are running outdated binaries.

No VM guests are running outdated hypervisor (qemu) binaries on this host.
Creating virtual environment '.DFetchVEnv'.
Virtual environment '.DFetchVEnv' created successfully.
Activated virtual environment: .DFetchVEnv
Checks passed...
Running DarwinFetch using python3
Traceback (most recent call last):
  File "/root/DarwinFetch/src/main.py", line 12, in <module>
    import py7zr
ModuleNotFoundError: No module named 'py7zr'
root@puma:~/DarwinFetch# 
```

**Comment by ak2766** [comment]
I removed a `clear` command in the script using `sed` so I could capture everything...

**Comment by ak2766** [comment]
I'll come back to this when I'm to install something other than Puma.  Right now though, I'm going to have to go hunt elsewhere as your script doesn't support downloading Puma install media, unless I missed something somewhere.

**Comment by ak2766** [comment]
Nevermind - I just noticed I need to enable some disabled sources...

**Comment by royalgraphx** [comment]
https://github.com/royalgraphx/DarwinFetch/blob/main/DarwinFetch.sh#L232-L257

Debian based installations had the following snippet removed:

```bash
# Check if all modules can be imported and handle failure if necessary
        check_pip
        check_requirements_import
        if [ $? -ne 0 ]; then
            install_requirements_pip
        fi
```

This was the reason, it was not ensuring modules can be loaded or was installing them, this was an oversight because DKVM does not inherently support Fedora or Debian, the distribution of choice is Arch Linux.

**Comment by ak2766** [comment]
Cool.  I was able to get around it by sourcing `activate` and running `pip install...` and now I'm downloading Puma.  Off to the races I go.

**Comment by ak2766** [comment]
Thanks for your efforts.

```
Config loaded successfully.
Available PowerPC Sources:
2. Mac OS X Leopard 10.5.6
11. Mac OS X Tiger 10.4.8
17. Mac OS X Panther 10.3.0
19. Mac OS X Jaguar 10.2.0
21. Mac OS X Puma 10.1.0
28. Mac OS X Cheetah 10.0.3
Enter the number of the source to download (or 'c' to cancel): 21

Selected Source: Mac OS X Puma 10.1.0 (5G640) - 000-000000

Initializing download of Apple Mac OS X 10.1 Puma.7z
100%|███████████████████████████████████████████████████████████████████████████████████| 486M/486M [00:46<00:00, 10.5MiB/s]

Download completed. File saved to: downloads/10.1.0_5G640/Apple Mac OS X 10.1 (10.1.5J34).7z
Unpacking files downloaded from sources...
Successfully unpacked: Apple Mac OS X 10.1 (10.1.5J34).7z
Press any key to continue...
```

**Comment by royalgraphx** [comment]
this has now been fixed via the recent commit [a7c7f63fe7656ca3499be3501fb74ce38dd557f6](https://github.com/royalgraphx/DarwinFetch/commit/a7c7f63fe7656ca3499be3501fb74ce38dd557f6)

Thank you for bringing this to my attention, not many use Debian for this, this would have flown under my radar for months
